### PR TITLE
Handle CLOSE reason as Incoming

### PR DESCRIPTION
**Pull request by dvdplm**
When a remote sends us a CLOSE message it may contain a reason code and "application data" with further details about why the connection was closed.

Before this PR there was no way for applications using `soketto` to know what the code/data was sent from the remote.

Further context [here](https://github.com/libp2p/rust-libp2p/issues/2021).

This is an alternative to #29 and returns the close reason as `Incoming::Closed` instead of `Error::Closed(reason)`.

### CODE REVIEW

File: src/data.rs [hunk 1]
```diff
@@ -48,11 +52,22 @@ impl Incoming<'_> {
...
+        }
+    }
+
     /// The length of data (number of bytes).
     pub fn len(&self) -> usize {
         match self {
             Incoming::Data(d) => d.len(),
-            Incoming::Pong(d) => d.len()
+            Incoming::Pong(d) => d.len(),
+            Incoming::Closed(reason) => std::mem::size_of_val(reason), // TODO: this makes no sense; use `unimplemented!()` here?
```
Comment (tomaka) [inline src/data.rs]: This method is indeed a bit questionable.
I don't see a situation where you would want the length of the payload of `Pong` and treat it the same way as the length of the payload of `Data`.

I'd go with removing the method entirely.

**Review by tomaka — APPROVED** [review]
Looks good to me after TODOs are fixed

**Comment by dvdplm** [comment]
@tomaka I think this is good to go now. The only relevant change since you last looked at this is https://github.com/paritytech/soketto/pull/31/commits/f5b90905f0e92c9850e7cc06a0adb35dc1e0de1f (don't bail early if we can't echo back their CLOSE e.g. if they already closed the connection).

**Review by tomaka — APPROVED** [review]
:+1:

**Review by maciejhirsz — APPROVED** [review]
LGMT, I don't know why we are now ignoring the errors on sending the ACK to close, but I also don't see how that would break anything.

**Comment by dvdplm** [comment]
> I don't know why we are now ignoring the errors on sending the ACK to close

Because otherwise we'd do an early exit and the code higher up the call chain would not see the `CloseReason`. The send errors stem from the other side having closed the connection already (which in my testing was pretty common).

**Review by niklasad1 — APPROVED** [review]

**Pull request merged**